packages/push/src/instruction/common/push_value.rs

```diff
@@ -0,0 +1,125 @@
+use std::fmt::Display;
+
+use super::super::{Instruction, instruction_error::PushInstructionError};
+use crate::{
+    error::{InstructionResult, MapInstructionError},
+    push_vm::HasStack,
+};
+
+/// An instruction that pushes a value type `T` on the
+/// top a stack of type `T` in the given state.
+///
+/// # Inputs
+///
+/// The `Push` instruction takes no inputs from the stacks.
+///
+/// # Behavior
+///
+/// The `Push` instruction clones its internal value and pushes on the stack of
+/// type `T`.
+///
+/// If the stack is full before this instruction is performed, then
+/// a fatal [`StackError::Overflow`](crate::push_vm::stack::StackError::Overflow) is returned,
+/// terminating the running of the program.
+///
+/// ## Action Table
+///
+/// The table below indicates the behavior in each of the different
+/// cases.
+///
+///    - The "`Stack<T>` full" column indicates whether the stack of type `T` is
+///      full before this instruction is performed.
+///    - The "Success" column indicates whether the instruction succeeds, and if
+///      not what kind of error is returned:
+///       - ✅: success (so two copies of the top block on the `Exec stack`)
+///       - ❗: recoverable error, with links to the error kind
+///       - ‼️: fatal error, with links to the error kind
+///    - The "Note" column briefly summarizes the action state in that case
+///
+/// | `Stack<T>` full |  Success | Note |
+/// | ------------- | ------------- | ------------- |
+/// | false | ✅ | The internal value is pushed onto the stack |
+/// | true | [‼️..](crate::push_vm::stack::StackError::Overflow) | Program is terminated |
+///
+/// # Errors
+///
+/// If the stack access returns any error other than a
+/// [`StackError::Underflow`](crate::push_vm::stack::StackError::Underflow)
+/// then this returns that as a [`Error::Fatal`](crate::error::Error::Fatal)
+/// error.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub struct PushValue<T>(pub T);
+
+impl<S, T> Instruction<S> for PushValue<T>
+where
+    S: Clone + HasStack<T>,
+    T: Clone,
+{
+    type Error = PushInstructionError;
+
+    fn perform(&self, state: S) -> InstructionResult<S, Self::Error> {
+        state.with_push(self.0.clone()).map_err_into()
+    }
+}
+
+impl<T> Display for PushValue<T>
+where
+    T: Display,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Push({})", self.0)
+    }
```

Comment:
Are we using [`derive_more`](https://docs.rs/derive_more`) in this project already? if we do just use the derive if we don't I'd keep this as is.

Reply:
A search suggests we are _not_ using `derive_more`. I know you've mentioned it a few times, but it doesn't look like we've ever brought it in.